README.md
# <p align="center"> MKlon3.5E
## <p align="center">Зарядное устройство на модулях ESP32DEVKITCV4, SAMD21 MINI и ILI9486 3.5".
## <p align="center">Проект Arduino, в разработке.
#### <p align="right">2023.12.16
### MKlon3.5E - это компактное ЗУ модульной конструкции с графическим интерфейсом и сенсорным управлением. Разработанное системное ПО позволяет реализовать пользовательские алгоритмы обслуживания аккумуляторных батарей. Проект расчитан на пользователей, обладающих навыками программирования в среде Arduino, а потому содержит лишь примеры реализации фрагментов алгоритмов пользователя с подробными комментариями. 

### Аппаратная часть проектировалась в среде Eagle v7.3.0, программная - в PlatformIO. 
<a name="menu"></a>
1. [Как это работает. Аппаратный аспект.](#hard)
2. [Как это работает. Программный аспект](#soft)
3. [Технические характеристики](#specifications)
4. [Тесты](#tests)
5. [Документы](#docs)
6. [Полезные ссылки](#useful_links)
7. [About Me](#about)

***
![Модель](https://github.com/olmoro/MKlon3.5v7/blob/main/documents/full/img/MKlon3.5full.png)
![Реализация](https://github.com/olmoro/MKlon3.5e/blob/main/documents/tests/MKlon3.5_Full.jpg)
![](https://github.com/olmoro/MKlon3.5e/blob/main/documents/img/MKlon3.5_all.png)
![](https://github.com/olmoro/MKlon3.5e/blob/main/documents/img/MKlon3.5_coo.png)
![](https://github.com/olmoro/MKlon3.5e/blob/main/documents/img/MKlon3.5_krk.png)

## <p align="center"> Как это работает.<a name="hard"></a>

Зарядное устройство состоит из двух блоков, каждый на печатной плате размером 100*65 мм. 
Межблочное соединение - высокий 16-контактный разъем.
![](https://github.com/olmoro/MKlon3.5e/blob/main/documents/img/MKlon3.5_cpu.png)
- Блок управления служит материнской платой для 38-контактного модуля [ESP32_DEVKITC_V4](https://aliexpress.ru/item/1005002443255665.html?spm=a2g2w.cart.cart_split.2.27194aa6uI5WfP&sku_id=12000020676652656&_ga=2.47565607.1828311252.1683559036-197161830.1681374836), 28-контактного  [SAMD21 M0-Mini](https://aliexpress.ru/item/1005002585114846.html?sku_id=12000021271636987&spm=a2g2w.productlist.search_results.0.58ef4aa657Lxkj), а также 3,5-дюймового дисплея [480*320 TFT ILI9486](https://aliexpress.ru/item/1005001999296476.html?spm=a2g2w.orderdetail.0.0.61a14aa6WcFlJS&sku_id=12000018365356569).

![](https://github.com/olmoro/MKlon3.5e/blob/main/documents/img/MKlon3.5_pow.png)
- Для силового блока "донором" послужил [XL4016 DC-DC Max 9A 300W](https://aliexpress.ru/item/32661166685.html?_ga=2.264151500.1168692661.1643047564-769654542.1642920280&sku_id=12000020944321893&spm=a2g39.orderlist.0.0.60754aa6RxSv8u), электролитические конденсаторы, шунты и дроссель которого были перенесены на плату прибора. В авторском варианте для подключения к источнику, аккумуляторной батарее и нагрузке использованы 6-амперные ноутбучные разъёмы. Второй вариант (он дешевле) предполагает использование двухконтактных разъёмов "под винт". 

[^](#menu)
***
## <p align="center">2. Как это работает. Программный аспект.<a name="soft"></a>
 Пользовательский интерфейс прибора реализован на ESP32, отлично приспособленном для этой задачи, а вот для чего он не приспособлен, так измерений с заданной погрешностью. Обычно это решается использованием INA226, но это всего лишь измеритель. Здесь же применен микроконтроллер на SAMD21, производящий не только измерения, причем с дифференциальными входами, но и математическую обработку, и генерацию сигнала для драйверов силового блока, сводя до разумного минимума время реакции на аварийные ситуации, возникающие в процессе работы. Связь между модулями производится по специальному протоколу через асинхронный интерфейс UART командами:
- ***txReadUIS()*** - Чтение напряжения, тока (мВ и мА) и двух байт состояния. Команда фоновая, запускается автоматически с периодом 0,1 секунды. Полученные данные выводятся на дисплей.
- ***txPowerAuto(spV, spI)*** - Запуск ПИД-регулятора в режиме заряда с заданными максимальными уровнями напряжения и тока.
- ***txPowerStop()*** - Команда прекратить заряд или разряд.
- ***txDischargeGo(spI)*** - Запуск ПИД-регулятора в режиме разряда с заданным максимальным уровнем тока.
- ***txVoltageAdj(spV)*** - Команда регулировки напряжения без остановки заряда.
- ***txCurrentAdj(spI)***  - Команда регулировки тока без остановки заряда.
- ***txDiscurrentAdj(spD)*** - Команда регулировки тока без остановки разряда.
- ***txSetPidCoeff(kp, ki, kd)*** - Команда ввода коэффициентов ПИД-регулятора.
- и ещё ***полсотни*** команд для настроек под аппаратную реализацию пользователя. ПО модуля разрабатывается в другом проекте, [здесь](https://github.com/olmoro/MKlon3.5e/blob/main/documents/fwMoroPid) файлы прошивки.

[^](#menu)
***
## <p align="center">3. Технические характеристики MKlon3.5e<a name="specifications"></a>
 Параметр | Ед | Значение | Погрешность |
------------------------------------:|----:|------------:|----------------------:|
|Зарядный ток Iз|А|0.05 – 6.0|±(0.005 Iз + 0.05)|
|Шаг установки зарядного тока|А| 0.01|
|Зарядное напряжение U|В|1.0 – 18.0|±(0.005 U + 0.05)|
|Шаг установки зарядного напряжения|В|0.01|
|Разрядный ток Iр|А|0.05 – 3.0|±(0.005 Iр + 0.05)|
|Шаг установки разрядного тока|А| 0.01|
|Максимальная рассеиваемая мощность|Вт|не менее 40|

- ***Источник питания*** - ноутбучный 19 В (4.74 ... 6.0 А)
- ***Интерфейс*** - USB, WiFi, BT
- ***Индикация*** - TFT 3.5 RGB, сенсорная панель 
- ***Диагностика*** - 2 разъема для логического анализатора
- ***Micro-SD карта*** - на плате дисплея
- ***Защиты*** - переполюсовка, перегрузка по току, КЗ

[^](#menu)
***
## <p align="center">4. Тесты.<a name="tests"></a>
- [Заряд АКБ Са-Са 60Ач, старт](https://github.com/olmoro/MKlon3.5e/blob/main/documents/tests/ch_60go.png)
- [Разряд максимальным током](https://github.com/olmoro/MKlon3.5e/blob/main/documents/tests/dis_type.png)
- [Работа системы теплоотвода](https://github.com/olmoro/MKlon3.5e/blob/main/documents/tests/Dis6A_t36.6.png)
- [Разряд минимальным током](https://github.com/olmoro/MKlon3.5e/blob/main/documents/tests/dis_min.png)
- В директории documents/tests есть 20-секундное видео.
 
[^](#menu)
***
## <p align="center">5. Документация.<a name="docs"></a>

- [Схемы](https://github.com/olmoro/MKlon3.5e/tree/main/documents/sch) 


[^](#menu)
***
## <p align="center">6.Полезные ссылки<a name="useful_links"></a>
- [MKlon3.5v7](https://github.com/olmoro/MKlon3.5v7)
- [OSS](https://ru.wikipedia.org/wiki/%D0%9E%D1%82%D0%BA%D1%80%D1%8B%D1%82%D0%BE%D0%B5_%D0%BF%D1%80%D0%BE%D0%B3%D1%80%D0%B0%D0%BC%D0%BC%D0%BD%D0%BE%D0%B5_%D0%BE%D0%B1%D0%B5%D1%81%D0%BF%D0%B5%D1%87%D0%B5%D0%BD%D0%B8%D0%B5)
- [Протокол обмена](http://leoniv.diod.club/articles/wake/wake.html)
- [Проекты на ESP32](https://randomnerdtutorials.com/projects-esp32/)
- [Как создать проект](https://dzen.ru/media/kotyara12/sozdanie-platformio--espidf-proekta-i-nastroika-platformioini-6324c68fb3d4c87d20ee80a6)
- [Как строятся графики](https://www.chartjs.org/)
- [Использование памяти в ESP32](https://dzen.ru/media/kotyara12/ispolzovanie-pamiati-v-esp32-63294a246a97da509e279eab)
- [Резистор в цепи затвора или как делать правильно](https://habr.com/en/articles/451152/)

[^](#menu)
***
## <p align="center">7. About Me<a name="about"></a>
🚀 I'm a full stack developer urk2t@yandex.ru

[^](#menu)

history.txt
.LOG

19:15 27.11.2023
======================= Проекты MKlon3.5E и MoroPidE =======================

Изменен формат команды 0x44 - передается без аргументов, очистка обоих регуляторов.

8:31 01.12.2023
Проба работы ПИД-регулятора на фиксированной частоте 1кГц
  par_mult         = 0x8000;
  par_max          = 0x0001;
  par_float_max    = 1.0f;

05.12.2023
txPowerOff()      0x29* Перевод в режим холостого хода (резерв,  )
txIdle()          0x2A* Перевод в режим холостого хода
RAM:   [=         ]   9.1% (used 48524 bytes from 532480 bytes)
Flash: [========= ]  93.9% (used 1231374 bytes from 1310720 bytes)

07.12.2023
devicefsm. : Удалены состояния
MAdjPidI
MPidFrequency
MLoadKpI  MLoadKiI  MLoadKdI

Переименовано:
MAdjPidV ->  MAdjPidVI  -> MAdjPidC (Charge)
kpV      ->  kpVI       -> kp

Новые:
  void ledStopGo();
  void autoStopGo(short spV, short spI);
  void autoIdleGo(short spV, short spI);

RAM:   [=         ]   9.1% (used 48500 bytes from 532480 bytes)
Flash: [========= ]  93.6% (used 1226790 bytes from 1310720 bytes)
                                 1226382

15.12.2023
Восстановление предыдущей версии 
Оставлена одна фиксированная частота пид-регулятора 1000hZ        - OK
RAM:   [=         ]   9.1% (used 48500 bytes from 532480 bytes)
Flash: [========= ]  93.6% (used 1226450 bytes from 1310720 bytes)

16.12.2023
Возврат предвычислений Ki и Kd на сторону ESP32 (-DKIKD)
  calkKpKiKd(p, i, d);
ESP32   - ok
SAMD21  - ok
RAM:   [=         ]   9.1% (used 48500 bytes from 532480 bytes)
Flash: [========= ]  93.6% (used 1226478 bytes from 1310720 bytes)

txAutoCurrentUp(short spV, short spI, short dI);   // 0x2B      - OK
di - дельта приращения тока (мА) за один шаг регулятора
на стороне samd не реализована

22.12.2023
Предвычисления ... функции преобразования коэффициентов:        - OK
  unsigned short kpToShort(float p);
  unsigned short kiToShort(float i);
  unsigned short kdToShort(float d); 
В NVS коэффициенты остаются в формате float, преобразования 
только при передаче на исполнение SAMD21

27.12.2023
Оптимизирован начальный загрузчик DOOT:
Добавлены инициализации PIDC и PIDD (0x43)      - OK
0x41      включает заряд??? - исправлено        - OK
Отредактирован project_config.h                 - OK
firmware_20231227.zip                           - OK


src/mtools.h
#ifndef _MTOOLS_H_
#define _MTOOLS_H_

#include "stdint.h"
  #include "project_config.h"

class MBoard;
class MDisplay;
class Preferences;

class MTools
{
  public:
    MTools(MBoard * board, MDisplay * display);
    ~MTools();
   
    //pointers are public - easier to access
    MBoard      * Board        = nullptr; // external
    MDisplay    * Display      = nullptr; // external
    Preferences * qPreferences = nullptr; // local

    // Буфер команд(ы) на исполнение, при необходимости будет очередь (queue) 
    uint8_t commBuff  = 0x00;   

    short error;              // Код ошибки обработки команды драйвером
    void  setErr(short err);
    short getErr();




    // Локализация
    bool getLocalization() const;
    void setLocalization(bool);
    
    // Диспетчер
    // short getModeSelection() const;
    // void  setModeSelection( short );
    void aboutMode( short modeSelection );

    void boot();


    // Флаг блокировки обмена с драйвером на время его рестарта
    bool getBlocking();
    void setBlocking(bool);

    // Флаг выбора режима коррекции смещения АЦП  nu?
    bool getTuningAdc();
    void setTuningAdc(bool);

    // Настройки АЦП
    short offsetAdc = 0x0000;
    short adcV      = 0x0000;
    short adcI      = 0x0000;

    short getAdcV();
    short getAdcI();
    
    void setSync( bool sync );
    bool getSync();

    // Переменные настроек измерителей MKlon2v7a 20230130
    unsigned short factorV   = MPrj::factor_v_default;   // Коэффициент преобразования в милливольты
    unsigned short smoothV   = MPrj::smooth_v_default;   // Коэффициент фильтрации
             short shiftV    = MPrj::shift_v_default;    // Приборное смещение в милливольтах

    unsigned short factorI   = MPrj::factor_i_default;   // Коэффициент преобразования в миллиамперы
    unsigned short smoothI   = MPrj::smooth_i_default;   // Коэффициент фильтрации
             short shiftI    = MPrj::shift_i_default;    // Приборное смещение в миллиамперах

  

    // ========== Управление ПИД-регулятором, частота фиксирована ==========

    // Pid parameters (значения остались от тестирования)
    uint16_t setpoint       = 0x0800;     // 
    uint16_t setpointU      = 0x3390;     // 13200
    uint16_t setpointI      = 0x0BB8;     //  3000
    short    deltaI         = 0x0000;
    uint16_t setpointD      = 0x0258;     //   600
    uint16_t ready          = 0x0000;     //  
    void  setReady(short ready);
    short getReady();

    unsigned short pidMode  = 0x01;       // 1 - начать с регулирования по напряжению
    unsigned short kp       = 0x0190;
    unsigned short ki       = 0x00C0;
    unsigned short kd       = 0x0190;
    unsigned short minOut   = 0x0220;
    unsigned short maxOut   = 0x1000;

    // Тестовые - для чтения SAMD
    unsigned short p       = 0;
    unsigned short i       = 0;
    unsigned short d       = 0;

    // PWM
    uint8_t  pwmInvert = (uint8_t)false;   // Выбор полярности PWM (v55: для отключения при сбросе - 0x00)
    uint16_t pwmPeriod = 0x1012;           // Выбор частоты (через период)

    uint8_t  swOnOff   = (uint8_t)false;

    void    setMilliVolt(short val);
    short   getMilliVolt();
    void    setMilliAmper(short val);
    short   getMilliAmper();

    // Текущие целочисленные в мВ и мА напряжение и ток преобразуются в вольты и амперы 
    void    setVoltageVolt(short);
    float   getVoltageVolt();
    void    setCurrentAmper(short);
    float   getCurrentAmper();
    void    setCelsius(short val);

    void    setState(unsigned short val);
    unsigned short getState();

    //  void  setProtErr(uint8_t val);  // protocol error - или подтверждения исполнения команды 

    float getRealVoltage(); //+m
    float getRealCurrent(); //+m


    unsigned short postpone = 0;      // Заданная задержка включения (отложенный старт), ч
    void  setPostpone(unsigned short hour);
    unsigned short getPostpone();

    bool getAP();

    // Чтение и запись NVS
    bool  readNvsBool  (const char * name, const char * key, const bool  defaultValue );
    short readNvsShort (const char * name, const char * key, const short defaultValue );
    int   readNvsInt   (const char * name, const char * key, const int   defaultValue );
    float readNvsFloat (const char * name, const char * key, const float defaultValue );

    void writeNvsBool  (const char * name, const char * key, const bool  bValue );
    void writeNvsShort (const char * name, const char * key, const short sValue );
    void writeNvsInt   (const char * name, const char * key, const int   iValue );
    void writeNvsFloat (const char * name, const char * key, const float fValue );

    float copyNvsFloat (const char * src, const char * dst, const char * key, const float defaultValue );
    float copyNvsFloat (const char * src, const char * dst, const char * key, const char * newkey, const float defaultValue );

    // 202207
    short updnInt(short value, short below, short above, short additives);
    float updnFloat(float value, float below, float above, float additives);

    bool  clearAllKeys (const char * name);
    void  removeKey(const char * name, const char * key);

    void clrTimeCounter();
    void clrChargeTimeCounter();
    void clrAhCharge();
    int  getTimeCounter();
    void setTimeCounter( int ivalue );

    uint8_t getBuffCmd();
    void    setBuffCmd(uint8_t cmd);

    short getCooler();
    void  setCooler(short val);

    // АЦП - настройки
    void  setAdcV(short val);
    void  setAdcI(short val);
    short getAdcOffset();
    void  setAdcOffset(short val);

    short getLtV();
    short getUpV();
    short getLtI();
    short getUpI();

    // ======================== ЦЕЛЕВЫЕ КОМАНДЫ УПРАВЛЕНИЯ ДРАЙВЕРОМ SAMD21 M0 MINI ========================

      // Команды чтения результатов измерений
    void txReadUIS();                                       // 0x10;
    void txGetU();                                          // 0x11 Чтение напряжения (мВ)
    void txGetI();                                          // 0x12 Чтение тока (мА)
    void txGetUI();                                         // 0x13 Чтение напряжения (мВ) и тока (мА)
    void txGetState();                                      // 0x14 Чтение состояния
    void txReady(short value);                              // 0x15* Согласование версий, готовность
      
      // Команды stop/go
    void txPowerAuto(short spV, short spI);                 // 0x20*
    void txPowerStop();                                     // 0x21*
    void txPowerMode(short spV, short spI, uint8_t mode);   // 0x22
    void txPowerVGo(short spV, short spI);                            // 0x22
    void txPowerIGo(short spV, short spI);                            // 0x22
    void txDischargeGo(short spD);                          // 0x24

    void txVoltageAdj(short spV);                           // 0x25
    void txCurrentAdj(short spI);                           // 0x26
    void txDiscurrentAdj(short spD);                        // 0x27

//    void txPowerOn();                                   // 0x28
    void txPowerOff();                                      // 0x29*
    void txIdle();                                          // 0x2A* - отменено
    void txAutoCurrentUp(short spV, short spI, short dI);   // 0x2B


      // Команды работы с измерителем напряжения
        // Множитель преобразования в милливольты
    void txGetFactorU();                                    // 0x30 Чтение
    void txSetFactorU(short val);                           // 0x31* Запись
    void txSetFactorDefaultU();                             // 0x32 Возврат к заводскому
        // Параметр сглаживания
    void txGetSmoothU();                                    // 0x33 Чтение
    void txSetSmoothU(short val);                           // 0x34* Запись
        // Приборный сдвиг
    void txGetShiftU();                                     // 0x35 Чтение
    void txSetShiftU(short val);                            // 0x36* Запись

      // Команды работы с измерителем тока
        // Множитель преобразования в миллиамперы
    void txGetFactorI();                                    // 0x38 Чтение
    void txSetFactorI(short val);                           // 0x39* Запись
    void txSetFactorDefaultI();                             // 0x3A Возврат к заводскому
        // Параметр сглаживания
    void txGetSmoothI();                                    // 0x3B Чтение
    void txSetSmoothI(short val);                           // 0x3C* Запись
        // Приборный сдвиг
    void txGetShiftI();                                     // 0x3D Чтение
    void txSetShiftI(short val);                            // 0x3E* Запись

    void txSetPidShortC();                            // 0x
    void txSetPidShortD();

      // Команды работы с ПИД-регулятором
    void txSetPidConfig(uint8_t m, float p, float i, float d, uint16_t minOut, uint16_t maxOut);   // 0x40 Запись

    void txSetPidCoeff(unsigned short m, float p, float i, float d);    // 0x41 Запись
    // void txSetPidCoeffV(float kp, float ki, float kd);      // 0x41* Запись
    // void txSetPidCoeffI(float kp, float ki, float kd);      // 0x41* Запись
    void txSetPidCoeffC(float p, float i, float d);      // 0x41* Запись
    void txSetPidCoeffD(float p, float i, float d);      // 0x41* Запись

    void txSetPidOutputRange(uint8_t m, uint16_t minOut, uint16_t maxOut);                               // 0x42
    void txSetPidReconfig(uint8_t m, float p, float i, float d,
                        uint16_t minOut, uint16_t maxOut);    // 0x43*, w/o clear
    void txPidClear();                                      // 0x44*

    void txGetPidTreaty();                                  // 0x47* Get shift, bits, hz

    //void txGetPidConfig();                                  // 0x48 get mode, kP, kI, kD, min, max - возвращает параметры текущего режима регулирования
    void txGetPidConfig(uint8_t m);                 // 0x48 (mode): kP, kI, kD режима по выбору


    void txSetPidChargeShort();               // 0x4C*
    void txSetPidDischargeShort();               // 0x4D*

    //void txSetPidTreaty(unsigned short shift, unsigned short bits, unsigned short hz);  // 0x4A Запись
  // #ifndef HZ1000
  //   void txSetPidFrequency(unsigned short hz);              // 0x4A* Запись
  // #endif  

  //   void txGetProbes();                                     // 0x50
  //   void txGetAdcOffset();                                  // 0x51
  //   void txSetAdcOffset(short val);                         // 0x52
  //   void txAdcAutoOffset();                                 // 0x53 (пока в резерве)


  //     // Команды тестовые (отменены?)
  //   // const uint8_t cmd_set_switch_pin            = 0x54; // sw_pin D4 PA14

  //   // const uint8_t cmd_set_power                 = 0x56; // пользоваться с осторожностью - выяснение пределов регулирования
  //   // const uint8_t cmd_set_discharge             = 0x57; // не проверена
  //   // const uint8_t cmd_set_voltage               = 0x58; // старая, не проверена
  //   // const uint8_t cmd_set_current               = 0x59; // старая, не проверена 
  //   // const uint8_t cmd_set_discurrent            = 0x5A; // старая, не проверена

  // //   void txSetCurrent(unsigned short val);     // 0x59  // 20231024

  //   //void txSetDiscurrent(uint8_t m, unsigned short val); // 0x5A      case MCmd::cmd_write_discurrent:          doSetDiscurrent();        break;  // 0x5A na
  //   void txSetDiscurrent(unsigned short val);     // 0x5A  // 20231022 
  //    //case MCmd::cmd_write_discurrent:          doSetDiscurrent();        break;  // 0x5A na

  //   // const uint8_t cmd_set_surge_compensation    = 0x5B; // параметры подавления всплеска напряжения na
  //   // const uint8_t cmd_set_idle_load             = 0x5C; // параметры доп.нагрузки ХХ

    // Команды задания порогов отключения
    void txGetLtV();                                        // 0x60
    void txSetLtV(short val);                               // 0x61
    void txSetLtDefaultV(short val);                        // 0x62
    void txGetUpV();                                        // 0x63
    void txSetUpV(short val);                               // 0x64
    void txSetUpDefaultV(short val);                        // 0x65

    void txGetLtI();                                        // 0x68
    void txSetLtI(short val);                               // 0x69
    void txSetLtDefaultI(short val);                        // 0x6A
    void txGetUpI();                                        // 0x6B
    void txSetUpI(short val);                               // 0x6C
    void txSetUpDefaultI(short val);                        // 0x6D

    // =====================================================================================================

    // Подсчет ампер-часов
    void  zeroAhCounter();                            // Обнуление счетчика ампер-часов заряда
    int   getChargeTimeCounter();
    float getAhCharge();
    void  chargeCalculations();

      // Подсчет задержки пуска
    bool  postponeCalculation();
  
    // ========================== FastPID ==========================
      /* Прототип: A fixed point PID controller with a 32-bit internal calculation pipeline.
        https://github.com/mike-matera/FastPID/tree/master/examples/VoltageRegulator

          Регулятор (SAMD21 MINI) оперирует с предвычисленными целочисленными данными типа unsigned short.
        Преобразование из float как и контроль корректности данных возлагается на центральный контроллер 
        ESP32. Коэффициенты kp, ki, kd должны быть преобразованы по формулам:
        kp * param_mult
        ki * param_mult / hz
        kd * param_mult * hz
        где param_mult - множитель, рассчитанный из соотношения согласованных между контроллерами
                         величин  param_shift и param_bits.

        По умолчанию задано:
        static constexpr uint8_t  param_shift = 8;
        static constexpr uint8_t  param_bits  = 16;

        static constexpr uint16_t param_max   = (((0x1ULL << param_bits)-1) >> param_shift);
        static constexpr uint16_t param_mult  = (((0x1ULL << param_bits)) >> (param_bits - param_shift));
        где param_max - максимум для проверки допустимости величин параметров.

        По умолчанию
        param_mult = 0x0100
        param_max  = 0x00FF
        Параметр hz в данной реализации всегда 10гц.  - отменено 2023 

        Интегральный член kp должен быть ненулевым.
      */ 

    // ============= Согласованные параметры обмена 20230217 =============
      // По умолчанию
    // static constexpr unsigned short param_shift = 8;  //9U; c20231107   =8
    // static constexpr unsigned short param_bits  = 16U;
    // static constexpr unsigned short param_max   = (((0x1ULL << param_bits)-1) >> param_shift);
    // static constexpr unsigned short param_mult  = (((0x1ULL << param_bits)) >> (param_bits - param_shift));
    static constexpr unsigned short param_hz    = 100U;
    // Из данных, полученных от драйвера, начальные - по умолчанию
//    unsigned short pMult = param_mult;      //_max;
//    unsigned short pMax  = param_max;
  //  static constexpr float fparh = (float)(param_max / param_mult);
    //unsigned short pHz   = param_hz;
    //unsigned short pidHz   = param_hz;
    short pidHz   = param_hz;
    
    unsigned short calkPHz(unsigned short hz);

  
    unsigned short getPHz();

  // ============= Регистр состояния драйвера =============

    static constexpr unsigned short status_switch           = 1U<<15; 
    static constexpr unsigned short status_power            = 1U<<14; 
    static constexpr unsigned short status_current_control  = 1U<<13; 
    static constexpr unsigned short status_voltage_control  = 1U<<12; 
    static constexpr unsigned short status_charge           = 1U<<11; 
    static constexpr unsigned short status_discharge        = 1U<<10; 
    static constexpr unsigned short status_idle             = 1U<<9; 
    static constexpr unsigned short status_pid              = 1U<<8;
    static constexpr unsigned short status_overheating      = 1U<<7; 
    static constexpr unsigned short status_overload         = 1U<<6; 
    static constexpr unsigned short status_power_limit      = 1U<<5; 
    static constexpr unsigned short status_reverse_polarity = 1U<<4; 
    static constexpr unsigned short status_short_circuit    = 1U<<3; 
    static constexpr unsigned short status_calibration      = 1U<<2; 
    static constexpr unsigned short status_upgrade          = 1U<<1; 
    static constexpr unsigned short status_reserve2         = 1U; 

    static constexpr unsigned short status_pid_voltage      = status_switch |
                                                              status_power |
                                                              status_voltage_control |
                                                              status_charge |
                                                              status_pid;   // учесть status_auto_mode

    static constexpr unsigned short status_pid_current      = status_switch |
                                                              status_power |
                                                              status_current_control |
                                                              status_charge |
                                                              status_pid;

    static constexpr unsigned short status_pid_discurrent   = status_switch |
                                                              status_discharge |
                                                              status_pid;

    unsigned short getStatusPidVoltage() {return status_pid_voltage;}
    unsigned short getStatusPidCurrent() {return status_pid_current;}
    //unsigned short getStatusPidAuto(){return status_pid_voltage | status_pid_current;}
    unsigned short getStatusPidDiscurrent() {return status_pid_discurrent;}


    // Дисплей 20230311

    void showAmp(float amp, uint8_t pls);
    void showAmp(float amp, uint8_t pls, short filtr);

    void showVolt(float volt, uint8_t pls);
    void showVolt(float volt, uint8_t pls, short filtr);

    //     ((Tools->getState() == Tools->getStatusPidVoltage()) |
    //  (Tools->getState() == Tools->getStatusPidCurrent())) ?  /*Green : Red*/
    //                                    Board->ledsGreen() : Board->ledsRed();
    void ledStopGo();

                // ((Tools->getState() == Tools->getStatusPidVoltage()) |
                //  (Tools->getState() == Tools->getStatusPidCurrent())) ?  /* 0x21 : 0x20 */
                //                                  Tools->txPowerStop() : Tools->txPowerAuto(spV, spI);
    void autoStopGo(short spV, short spI);
    void autoIdleGo(short spV, short spI);
           // (Tools->getState() == Tools->getStatusPidDiscurrent()) ?
            //                                   Tools->txPowerStop() : Tools->txDischargeGo(spD);

    void dischargeStopGo(short spD); 

// //    void calkKpKiKd(float p, float i, float d);
    // unsigned short kpToShort(float p);
    // unsigned short kiToShort(float i);
    // unsigned short kdToShort(float d);



  private:
    //==== PRIVATE ==== PRIVATE ==== PRIVATE ==== PRIVATE ==== PRIVATE ==== PRIVATE ==== PRIVATE ====

    // Пороги отключения по умолчанию: мВ, мА  (имитация аппаратной поддержки)
    // Согласовать с драйвером
    static constexpr short lt_default_v  =  -200;   // при переполюсовке
    static constexpr short up_default_v  = 19500;   // исполняется конструктивно
    static constexpr short lt_default_i  = -3050;   // максимальный ток разряда
    static constexpr short up_default_i  =  6050;   // максимальный ток заряда
      // Пороги отключения: мВ, мА  (имитация аппаратной поддержки)
    short ltV = lt_default_v;
    short upV = up_default_v;
    short ltI = lt_default_i;
    short upI = up_default_i;

    short milliVolt   = 0;        // Принятое от драйвера напряжение в милливольтах
    short milliAmper  = 0;        // Принятый от драйвера ток в миллиамперах

    float voltage     = 0.0f;    // Напряжение на клеммах аккумулятора, В
    float current     = 0.0f;    // Текущий измеренный ток, А

    unsigned short state = 0x0000;
    short celsius     = 0x0000;   // Температура радиатора в "попугаях" АЦП

      // Подсчет ампер-часов
    float ahCharge          = 0.0f;
    int   timeCounter       = 0;
    int   chargeTimeCounter = 0;



    //================= Measures =====================
    int   keyCode = 0;    

    short cool = 0;           // Скорость вентилятора - уточнить

    bool blocking;            // 
    bool tuningAdc  = false;  // Флаг подстройки смещения АЦП



    bool sync = false;                 // Флаг синхронизации параметров

};

#endif //_MTOOLS_H_
